packages/core/src/sentinel/basic-sentinel.ts

```diff
@@ -0,0 +1,132 @@
+
+const { fields, table } = convertToIdentifiers(SentinelActivities);
+
+/**
+ * A basic sentinel that blocks a user after 5 failed attempts in 1 hour.
+ *
+ * @see {@link BasicSentinel.supportedActions} for the list of supported actions.
+ */
+export default class BasicSentinel extends Sentinel {
+  /** The list of actions that are accepted to be reported to this sentinel. */
+  static supportedActions = Object.freeze([
+    SentinelActivityAction.Password,
+    SentinelActivityAction.VerificationCode,
+  ] as const);
+
+  /** The array of all supported actions in SQL format. */
+  static supportedActionArray = sql.array(
+    BasicSentinel.supportedActions,
+    SentinelActivities.fields.action
+  );
+
+  /**
+   * Asserts that the given action is supported by this sentinel.
+   *
+   * @throws {Error} If the action is not supported.
+   */
+  static assertAction(action: unknown): asserts action is SentinelActivityAction {
+    // eslint-disable-next-line no-restricted-syntax
+    if (!BasicSentinel.supportedActions.includes(action as SentinelActivityAction)) {
+      // Update to use the new error class later.
+      throw new Error(`Unsupported action: ${String(action)}`);
+    }
+  }
+
+  protected insertActivity = buildInsertIntoWithPool(this.pool)(SentinelActivities);
+
+  constructor(protected readonly pool: CommonQueryMethods) {
+    super();
+  }
+
+  /**
+   * Reports an activity to this sentinel. The sentinel will decide whether to block the user or
+   * not.
+   *
+   * Regardless of the decision, the activity will be recorded in the database.
+   *
+   * @param activity The activity to report.
+   * @returns The decision made by the sentinel.
+   * @throws {Error} If the action is not supported.
+   * @see {@link BasicSentinel.supportedActions} for the list of supported actions.
+   */
+  async reportActivity(activity: ActivityReport): Promise<SentinelDecisionTuple> {
+    BasicSentinel.assertAction(activity.action);
+
+    const [decision, decisionExpiresAt] = await this.decide(activity);
+
+    await this.insertActivity({
+      id: generateStandardId(),
+      ...activity,
+      decision,
+      decisionExpiresAt,
+    });
+
+    return [decision, decisionExpiresAt];
+  }
+
+  /**
+   * Checks whether the given target is blocked from performing actions.
+   *
+   * @returns The decision made by the sentinel, or `null` if the target is not blocked.
+   *
+   * @remarks
+   * All supported actions share the same pool of activities, i.e. once a user has failed to
+   * perform any of the supported actions for certain times, the user will be blocked from
+   * performing any of the supported actions.
+   */
+  protected async isBlocked(
+    query: Pick<SentinelActivity, 'targetType' | 'targetHash'>
+  ): Promise<Nullable<SentinelDecisionTuple>> {
+    const blocked = await this.pool.maybeOne<Pick<SentinelActivity, 'decisionExpiresAt'>>(sql`
+      select ${fields.decisionExpiresAt} from ${table}
```

Comment:
nit: put the sql in `quries` folder?

Reply:
i thought about this. on the other hand, `sentinel` is a relatively independent part, which means it's not the main business, in this way we can fully decouple from core in the future. (e.g. adaptive mfa can have a totally different package)

Reply:
comments added